tokenizers/src/models/unigram/trainer.rs

```diff
@@ -0,0 +1,685 @@
+use crate::models::unigram::{
+    lattice::Lattice,
+    model::Unigram,
+    unicode::{get_script, Script},
+};
+use crate::tokenizer::{AddedToken, Result, Trainer};
+use indicatif::{ProgressBar, ProgressStyle};
+use std::cmp::Reverse;
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+
+// A token and a score
+type SentencePiece = (String, f64);
+
+// A full sentence or word + it's count within the dataset
+type Sentence = (String, u32);
+
+fn digamma(mut x: f64) -> f64 {
+    let mut result = 0.0;
+    while x < 7.0 {
+        result -= 1.0 / x;
+        x += 1.0;
+    }
+    x -= 1.0 / 2.0;
+    let xx = 1.0 / x;
+    let xx2 = xx * xx;
+    let xx4 = xx2 * xx2;
+    result += x.ln() + (1.0 / 24.0) * xx2 - 7.0 / 960.0 * xx4 + (31.0 / 8064.0) * xx4 * xx2
+        - (127.0 / 30720.0) * xx4 * xx4;
+    result
+}
+
+fn to_log_prob(pieces: &mut [SentencePiece]) {
+    let sum: f64 = pieces.iter().map(|(_, score)| score).sum();
+    let logsum = sum.ln();
+    for (_, score) in pieces.iter_mut() {
+        *score = score.ln() - logsum;
+    }
+}
+
+/// A `UnigramTrainer` can train a `Unigram` model from `word_counts`.
+#[derive(Builder, Debug, Clone)]
+pub struct UnigramTrainer {
+    #[builder(default = "true")]
+    show_progress: bool,
+    #[builder(default = "8000")]
+    vocab_size: u32,
+    #[builder(default = "2")]
+    n_sub_iterations: u32,
+    #[builder(default = "0.75")]
+    shrinking_factor: f64,
+    #[builder(default = "vec![]")]
+    special_tokens: Vec<AddedToken>,
+
+    #[builder(default = "' '")]
+    space_char: char,
+
+    #[builder(default = "String::from(\"<unk>\")")]
+    unk_token: String,
+
+    #[builder(default = "false")]
+    treat_whitespace_as_suffix: bool,
```

Comment:
As we discussed, a lot of these options might be better suited as standalone `Normalizer` or `PreTokenizer`, and removed from there.